### PR TITLE
Combine `ty::Projection` and `ty::Opaque` into `ty::Alias`

### PR DESCRIPTION
**Pull request by compiler-errors**
Implements https://github.com/rust-lang/types-team/issues/79.

This PR consolidates `ty::Projection` and `ty::Opaque` into a single `ty::Alias`, with an `AliasKind` and `AliasTy` type (renamed from `ty::ProjectionTy`, which is the inner data of `ty::Projection`) defined as so:

```
enum AliasKind {
  Projection,
  Opaque,
}

struct AliasTy<'tcx> {
  def_id: DefId,
  substs: SubstsRef<'tcx>,
}
```

Since we don't have access to `TyCtxt` in type flags computation, and because repeatedly calling `DefKind` on the def-id is expensive, these two types are distinguished with `ty::AliasKind`, conveniently glob-imported into `ty::{Projection, Opaque}`. For example:

```diff
  match ty.kind() {
-   ty::Opaque(..) => 
+   ty::Alias(ty::Opaque, ..) => {}
    _ => {}
  }
```

This PR also consolidates match arms that treated `ty::Opaque` and `ty::Projection` identically.

r? @ghost

### CODE REVIEW

**Review by cjgillot — COMMENTED** [review]
I wonder what are the next steps. Can this be used to integrate in-trait opaques as a variant of `AliasKind`?

File: compiler/rustc_traits/src/chalk/lowering.rs [hunk 1]
```diff
@@ -66,10 +66,10 @@ impl<'tcx> LowerInto<'tcx, SubstsRef<'tcx>> for &chalk_ir::Substitution<RustInte
     }
 }
 
-impl<'tcx> LowerInto<'tcx, chalk_ir::AliasTy<RustInterner<'tcx>>> for ty::ProjectionTy<'tcx> {
+impl<'tcx> LowerInto<'tcx, chalk_ir::AliasTy<RustInterner<'tcx>>> for ty::AliasTy<'tcx> {
     fn lower_into(self, interner: RustInterner<'tcx>) -> chalk_ir::AliasTy<RustInterner<'tcx>> {
         chalk_ir::AliasTy::Projection(chalk_ir::ProjectionTy {
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_traits/src/chalk/lowering.rs:71]: Good point. I'll remove this impl.

**Comment by compiler-errors** [comment]
I think this is ready

r? types

File: compiler/rustc_hir_analysis/src/check/mod.rs [hunk 2]
```diff
@@ -352,11 +352,7 @@ fn bounds_from_generic_predicates<'tcx>(
         // insert the associated types where they correspond, but for now let's be "lazy" and
         // propose this instead of the following valid resugaring:
         // `T: Trait, Trait::Assoc = K` → `T: Trait<Assoc = K>`
-        where_clauses.push(format!(
-            "{} = {}",
-            tcx.def_path_str(p.projection_ty.item_def_id),
-            p.term,
-        ));
+        where_clauses.push(format!("{} = {}", tcx.def_path_str(p.projection_ty.def_id), p.term,));
```
Comment (oli-obk) [inline compiler/rustc_hir_analysis/src/check/mod.rs:355]:
```suggestion
        where_clauses.push(format!("{} = {}", tcx.def_path_str(p.projection_ty.def_id), p.term));
```

File: compiler/rustc_middle/src/ty/fast_reject.rs [hunk 3]
```diff
@@ -126,7 +126,7 @@ pub fn simplify_type<'tcx>(
             TreatParams::AsPlaceholder => Some(PlaceholderSimplifiedType),
             TreatParams::AsInfer => None,
         },
-        ty::Opaque(..) | ty::Projection(_) => match treat_params {
+        ty::Alias(ty::Opaque, ..) | ty::Alias(ty::Projection, _) => match treat_params {
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/fast_reject.rs:129]:
```suggestion
        ty::Alias(..) => match treat_params {
```

File: compiler/rustc_middle/src/ty/fast_reject.rs [hunk 4]
```diff
@@ -225,7 +225,10 @@ impl DeepRejectCtxt {
         match impl_ty.kind() {
             // Start by checking whether the type in the impl may unify with
             // pretty much everything. Just return `true` in that case.
-            ty::Param(_) | ty::Projection(_) | ty::Error(_) | ty::Opaque(..) => return true,
+            ty::Param(_)
+            | ty::Alias(ty::Projection, _)
+            | ty::Error(_)
+            | ty::Alias(ty::Opaque, ..) => return true,
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/fast_reject.rs:231]:
```suggestion
            | ty::Error(_)
            | ty::Alias(..) => return true,
```

File: compiler/rustc_middle/src/ty/fast_reject.rs [hunk 5]
```diff
@@ -344,7 +347,7 @@ impl DeepRejectCtxt {
             // projections can unify with other stuff.
             //
             // Looking forward to lazy normalization this is the safer strategy anyways.
-            ty::Projection(_) => true,
+            ty::Alias(ty::Projection, _) => true,
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/fast_reject.rs:350]:
```suggestion
            ty::Alias(..) => true,
```

File: compiler/rustc_middle/src/ty/fast_reject.rs [hunk 6]
```diff
@@ -323,7 +326,7 @@ impl DeepRejectCtxt {
                 _ => false,
             },
 
-            ty::Opaque(..) => true,
+            ty::Alias(ty::Opaque, ..) => true,
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/fast_reject.rs:329]:
```suggestion
```

File: compiler/rustc_trait_selection/src/traits/project.rs [hunk 7]
```diff
@@ -1379,8 +1379,12 @@ fn assemble_candidates_from_trait_def<'cx, 'tcx>(
...
     // If so, extract what we know from the trait and try to come up with a good answer.
     let bounds = match *obligation.predicate.self_ty().kind() {
-        ty::Projection(ref data) => tcx.bound_item_bounds(data.item_def_id).subst(tcx, data.substs),
-        ty::Opaque(def_id, substs) => tcx.bound_item_bounds(def_id).subst(tcx, substs),
+        ty::Alias(ty::Projection, ref data) => {
+            tcx.bound_item_bounds(data.def_id).subst(tcx, data.substs)
+        }
+        ty::Alias(ty::Opaque, ty::AliasTy { def_id, substs }) => {
+            tcx.bound_item_bounds(def_id).subst(tcx, substs)
+        }
```
Comment (oli-obk) [inline compiler/rustc_trait_selection/src/traits/project.rs:1387]:
```suggestion
        ty::Alias(_, ref data) => {
            tcx.bound_item_bounds(data.def_id).subst(tcx, data.substs)
        }
```

**Review by oli-obk — CHANGES_REQUESTED** [review]
r=me with some nits and a clean perf run

**Comment by compiler-errors** [comment]
bors is sleeping :zzz:

**Comment by compiler-errors** [comment]
:facepalm: github status page doesn't show anything wrong anymore, ugh.

**Comment by compiler-errors** [comment]
This looks like the same exact noisy crates that we saw in https://github.com/rust-lang/rust/pull/105350#issuecomment-1348348616

**Pull request merged**